tiledb/sm/cpp_api/capi_string.h

```diff
@@ -0,0 +1,108 @@
+/**
+ * @file   capi_string.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the C++ API for the TileDB CAPIString object.
+ */
+
+#ifndef TILEDB_CPP_API_CAPI_STRING_H
+#define TILEDB_CPP_API_CAPI_STRING_H
+
+#include <cassert>
+#include <optional>
+#include <stdexcept>
+#include "tiledb.h"
+
+namespace tiledb::impl {
+
+/**
+ * Manages the lifetime of a tiledb_string_t* handle and provides operations on
+ * it.
+ */
+class CAPIString {
+ public:
+  CAPIString(tiledb_string_t*&& handle) noexcept {
+    string_ = handle;
+    handle = nullptr;
+  }
+  ~CAPIString() {
+    if (string_ == nullptr) {
+      return;
+    }
+
+    capi_status_t result = tiledb_status(tiledb_string_free(&string_));
+    if (result != TILEDB_OK) {
+      log_warn("Could not free string; Error code: " + std::to_string(result));
+    }
+  }
+
+  // Disable copy and move.
```

Comment:
It's worth documenting that this class owns a resource. Thus copy construction and assignment can't be implemented, but move construction and assignment could be implemented but aren't because they're not needed. That way if anybody ever needs them, it'll be clear what to do.

Reply:
Done.